src/downloader.rs

```diff
@@ -7,18 +8,47 @@ use crate::warn;
 pub struct Downloader {
     client: reqwest::blocking::Client,
     tries: usize,
+    auth_map: HashMap<Option<String>, (String, Option<String>)>,
```

Comment:
So far, we can only use one username/password/host. Why make it a hashmap? Is that for future proofing or is it something you haven't implemented yet?

Reply:
The idea was for future-proofing and/or loading a list of authentication information from a file based on the conversation on the issue, but not sure if that's still in the scope of this issue. Happy to make it more simple if you think that makes more sense!

Reply:
If you see the need, then I don't see why not! It's just that right now it's lacking some way to use a different set of authentication tokens, or to register another one, so I just wanted some precisions regarding the use of the hashmap

Reply:
That’s fair, do you have any thoughts on the simplest way to parse these values from a file like was mentioned in the issue? To get around some of the issues you mentioned below about escape characters it seems like we’d want some kind of CSV parser, but not sure it would be ideal to add another dependency for this

Reply:
If the need for multiple authentications arises, how about using Serde and deserializing some JSON or YAML file? I don't see the "need" for it right now, but it could provide a way of escaping anything strange, and could be done ahead of time (before downloading I mean) using a fast and robust crate such as Serde

Reply:
Some other homemade format or CSV is also possible, but it would be harder imo for not much gain

Reply:
Gotcha, I could see that. @Skallwar do you have any thoughts on this? Wanted to check because you had originally mentioned it in the issue. Does a CSV-like format still make sense here, and is it still in the scope of this issue?

Reply:
I will give your work a look when I'm at home. For now I would say that first we should be able to take multiple logins from CLI. The file can wait (an other PR I guess :) )

Reply:
Sure! I'm basic this on the `reqwest` implementation of [`basic_auth`](https://docs.rs/reqwest/0.10.8/reqwest/struct.RequestBuilder.html#method.basic_auth) where username is required and password is optional.

I'm leaving the host as `Option<String>` with the assumption that when a host value isn't supplied the associated username and password should be the default authentication credentials for any hosts that don't otherwise match. Let me know if that seems off though!

Reply:
> I'm leaving the host as Option<String> with the assumption that when a host value isn't supplied the associated username and password should be the default authentication credentials for any hosts that don't otherwise match.

I will prefer not to give my username and password to every website which might not even need a login :laughing:. Making the host mandatory and using a regex to find which credential to send might be a better idea here

Reply:
Could we default the host to the host of the start URL supplied? That seems like it would be the default intended behavior, but I think not applying it to everything makes sense. And would we need to use regex if we're comparing against the host components of the URL?

Reply:
Thanks! I think I'm getting confused on the best way of incorporating globs though, since right now we're using a `HashMap` and unless I'm missing something it looks like we'd have to iterate through a `Vec` or slice of options to check the globs.

Is there a benefit to using a pattern to check against the host? It seems like different domains wouldn't require the same authentication credentials, and pulling the host string should handle the matching for us while reducing the amount of checks we need to do

Reply:
> it looks like we'd have to iterate through a Vec or slice of options to check the globs

Yes you are right. We could iterate on the given login every time we find a new host and cache the results in a hashtable

> It seems like different domains wouldn't require the same authentication credentials.

For example github.com and gist.github.com use the same credentials

Reply:
But yes, instead of future proofing, just go ahead with a basic version without globs it will be an awesome improvement anyway

Reply:
Sounds great! Should be implemented in the latest commit